"""
Exemple de client gRPC pour tester le service
"""
import grpc
from protos import emergency_pb2, emergency_pb2_grpc


def create_alert_example():
    """Exemple : Cr√©er une alerte"""
    channel = grpc.insecure_channel('localhost:50051')
    stub = emergency_pb2_grpc.EmergencyAlertServiceStub(channel)
    
    request = emergency_pb2.AlertRequest(
        type=emergency_pb2.FIRE,
        description="Incendie dans un immeuble r√©sidentiel, 3√®me √©tage",
        location=emergency_pb2.Location(
            latitude=48.8566,
            longitude=2.3522,
            address="25 Rue de la R√©publique",
            city="Paris",
            zone="Zone Centre"
        ),
        priority=emergency_pb2.CRITICAL,
        reporter_name="Marie Dubois",
        reporter_phone="+33612345678",
        affected_people=8
    )
    
    try:
        response = stub.CreateAlert(request)
        print(f"‚úÖ Alerte cr√©√©e: {response.alert_id}")
        print(f"   Statut: {emergency_pb2.AlertStatus.Name(response.status)}")
        print(f"   Priorit√©: {emergency_pb2.Priority.Name(response.priority)}")
    except grpc.RpcError as e:
        print(f"‚ùå Erreur: {e.code()}: {e.details()}")


def get_active_alerts_example():
    """Exemple : R√©cup√©rer alertes actives"""
    channel = grpc.insecure_channel('localhost:50051')
    stub = emergency_pb2_grpc.EmergencyAlertServiceStub(channel)
    
    request = emergency_pb2.ZoneRequest(
        zone="Zone Centre",
        min_priority=emergency_pb2.HIGH
    )
    
    try:
        response = stub.GetActiveAlerts(request)
        print(f"üìã {response.total_count} alertes actives trouv√©es:")
        
        for alert in response.alerts:
            print(f"\n   ID: {alert.alert_id}")
            print(f"   Type: {emergency_pb2.AlertType.Name(alert.type)}")
            print(f"   Description: {alert.description[:50]}...")
            print(f"   Priorit√©: {emergency_pb2.Priority.Name(alert.priority)}")
    except grpc.RpcError as e:
        print(f"‚ùå Erreur: {e.code()}: {e.details()}")


def update_alert_status_example(alert_id):
    """Exemple : Mettre √† jour le statut"""
    channel = grpc.insecure_channel('localhost:50051')
    stub = emergency_pb2_grpc.EmergencyAlertServiceStub(channel)
    
    request = emergency_pb2.StatusUpdateRequest(
        alert_id=alert_id,
        new_status=emergency_pb2.IN_PROGRESS,
        assigned_team="Pompiers Caserne 5",
        notes="√âquipe de 6 pompiers en route, ETA 8 minutes"
    )
    
    try:
        response = stub.UpdateAlertStatus(request)
        print(f"‚úÖ Alerte mise √† jour: {response.alert_id}")
        print(f"   Nouveau statut: {emergency_pb2.AlertStatus.Name(response.status)}")
        print(f"   √âquipe: {response.assigned_team}")
    except grpc.RpcError as e:
        print(f"‚ùå Erreur: {e.code()}: {e.details()}")


def subscribe_alerts_example():
    """Exemple : S'abonner aux alertes en streaming"""
    channel = grpc.insecure_channel('localhost:50051')
    stub = emergency_pb2_grpc.EmergencyAlertServiceStub(channel)
    
    request = emergency_pb2.SubscribeRequest(
        zones=["Zone Centre", "Zone Nord"],
        types=[emergency_pb2.FIRE, emergency_pb2.MEDICAL_EMERGENCY],
        min_priority=emergency_pb2.HIGH
    )
    
    print("üîÑ Streaming des alertes en temps r√©el...")
    print("   (Appuyez sur Ctrl+C pour arr√™ter)\n")
    
    try:
        for alert in stub.SubscribeAlerts(request):
            print(f"üö® NOUVELLE ALERTE:")
            print(f"   ID: {alert.alert_id}")
            print(f"   Type: {emergency_pb2.AlertType.Name(alert.type)}")
            print(f"   Zone: {alert.location.zone}")
            print(f"   Priorit√©: {emergency_pb2.Priority.Name(alert.priority)}")
            print(f"   Description: {alert.description[:80]}...")
            print()
    except KeyboardInterrupt:
        print("\n‚úã Streaming arr√™t√©")
    except grpc.RpcError as e:
        print(f"‚ùå Erreur: {e.code()}: {e.details()}")


def health_check_example():
    """Exemple : Health check"""
    channel = grpc.insecure_channel('localhost:50051')
    stub = emergency_pb2_grpc.EmergencyAlertServiceStub(channel)
    
    request = emergency_pb2.HealthCheckRequest()
    
    try:
        response = stub.HealthCheck(request)
        print(f"üíö Service: {response.status}")
        print(f"   Version: {response.version}")
        print(f"   Alertes actives: {response.active_alerts}")
        print(f"   Subscribers: {response.subscribers}")
    except grpc.RpcError as e:
        print(f"‚ùå Service unhealthy: {e.code()}")


if __name__ == '__main__':
    print("=" * 60)
    print("üè• Emergency Alert gRPC Client Examples")
    print("=" * 60)
    
    # 1. Health check
    print("\n1Ô∏è‚É£ Health Check:")
    health_check_example()
    
    # 2. Cr√©er une alerte
    print("\n2Ô∏è‚É£ Cr√©er une alerte:")
    create_alert_example()
    
    # 3. R√©cup√©rer alertes actives
    print("\n3Ô∏è‚É£ R√©cup√©rer alertes actives:")
    get_active_alerts_example()
    
    # 4. S'abonner (streaming)
    # D√©commenter pour tester le streaming
    print("\n4Ô∏è‚É£ S'abonner aux alertes:")
    subscribe_alerts_example()